Switch from bower to npm and update JS libraries

diosmosis opened the pull request:
Still a WIP.

NOTE: This will require updates to premium plugins.

TODO (in other PRs most likely):
* [x] get tests to pass
* [ ] update individual libraries and get tests to pass
* [x] BC for bower_components folder
* [ ] run tests of premium plugins and get them to pass
* [ ] replace other libraries in libs/ and in plugin directories

Refs #12961 
Fixes #12067

diosmosis committed:
Starting to replace bower with npm: use updated jquery-ui from npm.

diosmosis committed:
Use jquery from npm./

diosmosis committed:
Use angular from npm.

diosmosis committed:
Add more angular libraries.

diosmosis committed:
Use chroma-js from npm.

diosmosis committed:
Remove html5shiv, does not appear to be used.

diosmosis committed:
Use iframe resizer library from npm.

diosmosis committed:
Add jquery-mousewheel

diosmosis committed:
Add jquery.dotdotdot from npm.

diosmosis committed:
Get jquery.scrollto from npm.

diosmosis committed:
Get jscrollpane from npm.

diosmosis committed:
Get materialize-css from npm.

diosmosis committed:
Get mousetrap from npm.

diosmosis committed:
Remove ngDialog, as it is no longer used.

diosmosis committed:
Install qrcode.js from npm.

diosmosis committed:
Get sprintf-js from npm.

diosmosis committed:
Get visibillityjs from npm.

diosmosis committed:
Remove bower.json.

diosmosis committed:
Getting parts of matomo to work w/ new versions of libraries installed by npm.

diosmosis committed:
Ignore node_modules subfolders in node_modules since we only use npm for frontend dependencies + fix a test.

diosmosis committed:
Fix button padding.

diosmosis committed:
Fix series picker checkboxes.

diosmosis committed:
Fix karma conf.

diosmosis committed:
Fix CSS in right place.

diosmosis committed:
Fixing more css/less issues.

diosmosis committed:
More test fixes.

Findus23 commented:
@diosmosis Thanks for starting this work
See also #12067 for an old example package.json

It also seems like this doesn't really solve #12961 as jQuery <3 is still used.

diosmosis commented:
Currently just getting tests to pass w/ npm libs, some of whom required an update (eg, materializecss < 1.0 includes some node dependencies which we can't include in the repo, & some repos don't have the versions we use available). Then I'll upgrade libs and get tests to pass again.

Thanks for the warning about dotdotdot!

tsteur commented:
@diosmosis fyi changing the bower_components folder will be a BC break. They are used in various of our extra plugins and premium features. I assume also other plugins use it. Ideally we would keep the same directory structure and change it in Matomo 4 if possible. 

Also just to confirm we still commit the files to the repository right? Just making sure as it really does make installing and updating easier as you wouldn't need node.js to install it and makes sure everyone is using the same version. I would even go so far and commit the vendor directory eventually. Personally, I rarely run composer install after a branch change and things might not always be tested properly because of this. Just makes things easier and more consistent.

diosmosis commented:
> fyi changing the bower_components folder will be a BC break. They are used in various of our extra plugins and premium features. I assume also other plugins use it. Ideally we would keep the same directory structure and change it in Matomo 4 if possible.

Unfortunately, npm doesn't support specifying a different node_modules folder (though yarn does, so npm might in the future). I think for this we'd have to either change premium plugins to use the correct folder based on the matomo version, or give up until Matomo 4.

> Also just to confirm we still commit the files to the repository right?

Yes, so far no complications. If a library decides to include dependencies then it could get annoying, but I don't think it would be a real blocker.

tsteur commented:
> Unfortunately, npm doesn't support specifying a different node_modules folder (though yarn does, so npm might in the future). I think for this we'd have to either change premium plugins to use the correct folder based on the matomo version, or give up until Matomo 4.

Was thinking a symlink but reckon that might not be working on all systems. I reckon we could otherwise - if needed - have a little to do the npm update and manually move `node_modules` to `bower_components` as we commit it anyway? may be bit too annoying though.

It's likely not only premium features that use it but all plugins. Ideally we check the plugins on the marketplace to see if any of them use it (I think we used to have a search engine for that but don'T remember URL or if it still works). Most plugins won't be on the marketplace though...

Actually just thought of a better solution: We can rewrite the path in the asset manager and it'll be no problem :)

diosmosis commented:
> I reckon we could otherwise - if needed - have a little to do the npm update and manually move `node_modules` to `bower_components` as we commit it anyway? may be bit too annoying though.

This could be put into a command and a test added to ReleaseCheckListTest (since we have the list of files we use in the getJavaScriptFiles hook).

> Actually just thought of a better solution: We can rewrite the path in the asset manager and it'll be no problem :)

I see you beat me to it :) I was thinking of creating symlinks when merging assets, but using a map of bower_components => node_modules links works better. Will do this.

diosmosis committed:
Limit selection style fix.

diosmosis committed:
Fix more tests + materializecss issues.

diosmosis committed:
Fix a couple more tests.

diosmosis committed:
Fix annotation styling.

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis commented:
Moved this to 3.11 since it should probably be in a beta.

Findus23 commented:
I agree. As much as I am for updating all JS libraries, this might introduce quite a few subtle bugs and we should especially get as many people as possible to beta-test this.

diosmosis commented:
Moving to 4.0 as discussed in #12961

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
remove error

diosmosis committed:
add ng-dialog files

diosmosis committed:
try to fix some tests

diosmosis committed:
Fix several UI tests.

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
fixing more build issues

diosmosis committed:
Fix several more tests and issues.

diosmosis committed:
fixing more tests

diosmosis committed:
split ui tests into 3 groups

diosmosis committed:
fix several more issues and tests

diosmosis committed:
Fix some console warnings on chrome.

diosmosis committed:
Updating more test files.

diosmosis committed:
fix some tests

diosmosis committed:
yet more fixes

diosmosis committed:
couple more fixes

diosmosis committed:
another form fix

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
Fix some tests.

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
update screenshot

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
update more expected files

diosmosis committed:
Fix two more form issues.

diosmosis committed:
Merge branch '3.x-dev' into 12961-npm-and-updates

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
test commit (travis is not fetching submodules for some reason...)

diosmosis committed:
remove scripts removed in merge

diosmosis committed:
three travis builds

diosmosis committed:
split UIIntegrationTest into multiple suites

diosmosis committed:
lets try this again

diosmosis committed:
updae several expected screenshots

diosmosis committed:
fix some more tests

diosmosis committed:
fix plugin details material select and tabs initiaialization

diosmosis committed:
update screenshots + css fix

diosmosis committed:
re-initialize materialize tabs since we load jquery-ui afterwards for datepicker which unsets materialize tabs

diosmosis committed:
update more screenshots and fix couple more issues

diosmosis committed:
update more screenshots + tweak to test

diosmosis committed:
more fixes and screenshot updates

diosmosis committed:
fix some issues and update more screenshots

diosmosis committed:
update submodules

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
more fixes + more updated screenshots

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
more css fixes and test fixes

diosmosis committed:
couple fixes & updated screenshots

diosmosis committed:
update screenshots

diosmosis committed:
fix random failure

diosmosis committed:
Map old bower_component files to new ones in AssetManager.

diosmosis committed:
Remove node_modules from PR and ignore node_modules in .gitignore and remove libs/jquery (adding new required libs).

diosmosis committed:
Update .travis.yml.

diosmosis committed:
update screenshots

diosmosis committed:
fixing some more issues

diosmosis committed:
fix copy pasta

diosmosis committed:
update screenshots

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
fix more modal issues

diosmosis committed:
Update css & screenshots.

diosmosis committed:
correct close modal calls

diosmosis committed:
fixing more issues

diosmosis committed:
data-target instead of data-activates

diosmosis committed:
fix more tests

diosmosis committed:
trigger build that works?

diosmosis committed:
fix more tests

diosmosis committed:
update submodule

diosmosis committed:
debugging travis

diosmosis committed:
more debugging

diosmosis committed:
try to fix test

diosmosis committed:
fix modal no button click in test

diosmosis committed:
update more screenshots

diosmosis committed:
couple more test fixes

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
some more fixes and updated screenshots

diosmosis committed:
update screenshots

diosmosis commented:
@tsteur / @sgiehl this should be ready for a review. This PR was meant to just be a switch to npm, but some libraries we use don't have the versions we use in npm, so it included several upgrades to the minimum version on npm (including materialize & jquery-ui). So after it's merged, another PR will have to be done which runs npm update. There are also some libraries in libs/ I didn't replace, but probably could be.

sgiehl reviewed (changes requested):
Roughly looked through the code changes and left a few comments.
Will now do some local testing and add some more comments if anything is broken.
Btw. there are some smaller changes in colors and margins. Doesn't look bad to me, so should imho be fine.
~Also the changes to `plugins/Overlay/tests/UI/expected-screenshots/Overlay_page_link_clicks.png` doesn't seem correct. The test is failing on 4.x-dev. Not sure why that part isn't shown anymore.~ see #16016

sgiehl commented:
Visitor profile doesn't opened centered anymore, and there are some pixels overlapping on the grey headers:
![image](https://user-images.githubusercontent.com/1579355/82422667-07c58580-9a83-11ea-9831-f03de335e590.png)

All other popovers are centered when they are opened, but are not repositioned anymore after the content was loaded.

sgiehl commented:
Hovering row actions on subtable rows now show both tooltips:

![image](https://user-images.githubusercontent.com/1579355/82423444-0a74aa80-9a84-11ea-889a-bc50a0bad47a.png)

sgiehl commented:
Browsing the marketplace though normal menu looks bit broken (looks better when visiting though admin menu):
![image](https://user-images.githubusercontent.com/1579355/82424207-1c0a8200-9a85-11ea-8ed2-15111fbab7a9.png)
Also the button text color was white for disabled buttons before.

sgiehl commented:
Drop down for compare periods looks broken:
![image](https://user-images.githubusercontent.com/1579355/82424457-80c5dc80-9a85-11ea-9204-a006c343f43d.png)

sgiehl commented:
spacing between combined input fields in CustomAlerts plugin is broken:

![image](https://user-images.githubusercontent.com/1579355/82430951-50367080-9a8e-11ea-8835-92102ed55dbf.png)

sgiehl commented:
GoTo links on tracking code page don't work anymore:

![image](https://user-images.githubusercontent.com/1579355/82431617-26317e00-9a8f-11ea-8d87-3c0b406b1a2c.png)

sgiehl commented:
@diosmosis that's all I found so far by clicking through the UI. There might be some more regressions maybe, but imho it should be fine to merge this once those I already found are fixed. We can fix anything else later to avoid more merge conflicts with this PR

diosmosis commented on libs/HTML/QuickForm2/Element/InputCheckable.php:
```diff
@@ -153,9 +153,7 @@ public function __toString()
-            $label = '<label for="' . htmlspecialchars(
-                         $this->getId(), ENT_QUOTES, self::getOption('charset')
-                     ) . '">' . $this->data['content'] . '</label>';
+            return '<label>' . parent::__toString() . '<span>' . $this->data['content'] . '</span></label>';
```
Only way it works w/ new materialize is w/ the input nested in the label.

diosmosis commented on package.json:
```diff
@@ -0,0 +1,50 @@
+    "jquery.browser": "^0.1.0",
+    "jquery.dotdotdot": "^1.8.3",
+    "jquery.scrollto": "^1.4.14",
+    "jscrollpane": "^2.2.1",
```
👍

sgiehl replied:
package-lock.json might still need an update

diosmosis commented on plugins/CoreHome/angularjs/period-selector/period-selector.directive.html:
```diff
@@ -68,15 +66,22 @@ <h6>{{ 'General_Period'|translate }}</h6>
+                <span>Compare to:</span>
+            </label>
 
+            <!-- TODO: missing some translations -->
```
👍

diosmosis commented on plugins/CoreHome/javascripts/materialize-bc.js:
```diff
@@ -0,0 +1,18 @@
+/*!
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
```
👍

diosmosis commented on plugins/CorePluginsAdmin/angularjs/form-field/field-field-array.html:
```diff
@@ -1,8 +1,8 @@
+    <label for="{{ formField.name }}" ng-bind-html="formField.title"></label>
     <div matomo-field-array
-         name="{{ formField.name }}"
+         name="formField.name"
```
It's because we're binding a value to the matomo-field-array directive rather than setting the name attribute.

diosmosis committed:
apply some review feedback

diosmosis committed:
Fix modal centering and initial top position.

diosmosis commented:
> Hovering row actions on subtable rows now show both tooltips:

I'm not sure how best to fix this one. It's due to the tooltips being nested, only way around it I can see is to add the tooltip to the label column. However if there's no cell tooltip, then it won't display over other columns...

diosmosis committed:
apply some more pr feedback

diosmosis committed:
another pr fix

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
update submodule

diosmosis committed:
remove style

sgiehl commented:
Another note that came up to my mind: In the original issue we discussed if it might make sense to add the node_module files to git, so they are included in the checkout. Otherwise guess we need to adjust some guides about how using Matomo from git and also adjust the build script, as the files need to be included in the release package

sgiehl committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

sgiehl committed:
fix color of checkbox mark

sgiehl committed:
Fix centering the popover

Note: centering the popover relative to body doesn't work poperly anymore, maybe because of the fixed position of the widget overlay. Using the ui-widget-overlay to position the dialog centered seems to work

sgiehl committed:
ignore line endings check for node modules

sgiehl committed:
updates expected UI files

sgiehl committed:
submodule updates

sgiehl committed:
use minified file of jquery.browser.js

sgiehl commented:
@diosmosis I've updated branch and also fixed some of the tiny issues I had mentioned before.
Did you already have time to look at the tooltip issue? Should imho be the last remaining one...

diosmosis commented:
@sgiehl the fix for the tooltip issue is to switch to materialize, which I haven't looked into yet. The archiving PRs are currently my focus.

sgiehl committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

sgiehl committed:
updates expected test files

sgiehl committed:
ensure to hide parent tooltips when row action tooltips are shown

sgiehl commented:
@diosmosis I've tried switching to materialize tooltips. But it's a bit more work to do as I thought. We are using callbacks in some cases to determine the content of the tooltips and whether they should be shown or not. That is not supported by materialize tooltips. So we would need to change that. Also materialize uses the same jQuery wrapper for tooltips as jquery ui, so we can't use the `.tooltip()` method anymore or would need to use a custom jquery ui build without the tooltips...

I've now pushed a hackish solution that only fixes the issue for the datatable row actions. Might be enough for the moment, as I didn't find any other overlapping tooltip yet. Maybe we can go with this solution for now and create a follow up issue to maybe even remove jquery ui completely with a later release.

diosmosis commented:
> I've now pushed a hackish solution that only fixes the issue for the datatable row actions. 

If it works, this sounds good to me, whatever takes the least amount of time at this point.

diosmosis committed:
tweak less for datatable navigation

diosmosis committed:
style is only useful for dashboard

diosmosis committed:
update some screenshots

diosmosis committed:
Merge branch '4.x-dev' into 12961-npm-and-updates

diosmosis committed:
update submodules
